File: python/euler/bin/p0090.py
# project euler: problem 90

from itertools import combinations, combinations_with_replacement

# For simplicity, I have replaced all 9's with 6.
SQUARES = [
    (0, 1),  # 1^2
    (0, 4),  # 2^2
    (0, 6),  # 3^2
    (1, 6),  # 4^2
    (2, 5),  # 5^2
    (3, 6),  # 6^2
    (4, 6),  # 7^2, 8^2
    (8, 1),  # 9^2
]


def is_contain(two_dice: tuple[set[int], set[int]], sq: tuple[int, int]) -> bool:
    if sq[0] in two_dice[0] and sq[1] in two_dice[1]:
        return True
    if sq[0] in two_dice[1] and sq[1] in two_dice[0]:
        return True

    return False


def compute() -> str:
    cnt = 0
    all_dice = map(lambda x: set(x), combinations([0, 1, 2, 3, 4, 5, 6, 7, 8, 6], 6))
    for two_dice in combinations_with_replacement(all_dice, 2):
        if all(is_contain(two_dice, sq) for sq in SQUARES):
            cnt += 1

    return str(cnt)


def solve() -> str:
    return compute()

File: typescript/euler/bin/p0090.ts
// project euler: problem 90

import {
  combinations,
  combinationsWithReplacement,
} from "combinatorics/mod.ts";

const check_squares = (dice: number[][]): boolean => {
  // deno-fmt-ignore
  const squares = [[0, 1], [0, 4], [0, 6], [1, 6], [2, 5], [3, 6], [4, 6], [8, 1]];

  const is_contained = (x: number[]) => {
    if (dice[0].includes(x[0]) && dice[1].includes(x[1])) {
      return true;
    }
    if (dice[0].includes(x[1]) && dice[1].includes(x[0])) {
      return true;
    }

    return false;
  };

  return squares.every(is_contained);
};

export const compute = (): string => {
  let acc = 0;
  for (
    const two_dice of combinationsWithReplacement(
      combinations([0, 1, 2, 3, 4, 5, 6, 7, 8, 6], 6),
      2,
    )
  ) {
    if (check_squares(two_dice)) {
      acc += 1;
    }
  }

  return String(acc);
};

export const solve = (): string => compute();
